### ISSUE TITLE
Feature Request: support IORedis (to support sentinels and clusters)

### ISSUE THREAD

**Issue by bgrieder** [capriza/node-busmq#22]
Although [node-redis](https://github.com/NodeRedis/node_redis) and [IORedis](https://github.com/luin/ioredis) have planned to ["merge"](https://github.com/luin/ioredis#consolidation-its-time-for-celebration), it has not happened yet and only IORedis supports sentinels and clusters out of the box.

Have you already tried injecting IORedis instead of node-redis as a client ?

Is the only modification required in `connection.js` ? (e.g. replace the `createClient` call with the corresponding call for IORedis)

**Comment by fujifish** [comment]
I actually did briefly look into that option a while back but found that there are some fundamental differences so it wasn't a drop-in replacement, such as pipelining and the multi command.
Basically it requires a modification in connection.js, but more more than just createClient.

I'd be happy to get a pull request if you're willing to invest in it.

**Comment by bgrieder** [comment]
@fujifish I have started hacking it and, yes, it is no trivial for sure !

One test I am struggling with is `producer attach -> producer push -> consumer attach -> consumer receive`. Since this is based on pub/sub and the consumer `subscribes` after the producer `publishes`, I do not understand how you expect this to work.
I realize the test passes with node-redis but does not if you simulate it using redis-cli (and ioredis as a matter of fact); `SUBSCRIBE` will not give you messages published *before* you subscribe.

Is this a realistic scenario ? Am I missing something ?

**Comment by fujifish** [comment]
first of all thanks for giving it a shot.

busmq uses queues for message delivery, pub/sub is used only for notifying any _current_ consumers that they can get a message from the queue. If a message is pushed to a queue, it will remain there until some consumer attaches to the queue, at which point it should start getting the messages, even those that were pushed _before_ it attached. That's the whole point!
In short, yes, it's very realistic.

The test you mentioned checks exactly that use case, which is important for a reliable message bus.

For use cases where it's not important to store messages if there are no consumers, pure pub/sub can be used.

**Comment by bgrieder** [comment]
@fujifish Makes sense. Thanks. 
I actually found what it was. The code "forces" an initial consumption of existing messages when the consumer subscribes. This initial consumption relies on the connection emitting a 'subscribe' event. IORedis does not emit that event. Emitting it on its behalf, fixed the issue.

**Comment by bgrieder** [comment]
@fujifish All tests are now passing with IORedis (and node-redis)
My code relies on the environment variable BUSMQ_USE_IOREDIS to be set to ether `true` or `yes` to use the ioredis library instead of node-redis.

As it is, there is no added functionality using IORedis instead of node-redis

What would you prefer:

1- I create a first pull request, you review that and merge it if you find it appropriate (I also hope you can give IORedis a shot in a more production like scenario and maybe run a couple of benchmarks)

2- You wait for me to incorporate sentinel and cluster support before creating the pull request

**Comment by fujifish** [comment]
Nice, that was pretty fast. I prefer option 1.

**Comment by bgrieder** [comment]
I estimate I spent about 10 hours on it.  IoRedis documentation is most certainly not the best part about it. 

I am away from my desk right now.  If you are in a hurry to evaluate you can check my fork of your library.  I have all my changes committed.

**Comment by fujifish** [comment]
No hurry, I'll take a look at your fork

**Comment by fujifish** [comment]
I reviewed the code, it seems simple enough.

1. I'm not a big fan of environment variables so I'll probably change that after I pull
2. ioredis does not support the 'subscribe' event so need to emulate it, however you only subscribe to channel[0], but emit the subscribe event to all channels which isn;t very clean. I think it's safe to change the function to just handle the one channel instead of an array of channels so the code is tighter.
3. For some time now I've been wanting to change the queue.push method to use the lua script to perform all of what the multi/exec does. This seems like a good opportunity so that there won't be any differences between the two libraries. I can do that after I pull.
4. Please remove the dump.rdp file and the entire node_module/ directory you (probably accidently) committed.

Can you give it another go?

**Comment by bgrieder** [comment]
1. Sure. Let me know what you went for (constructor param I guess ?)
2. Good catch. The [0] index is wrong; that was for a test. I removed it.
3. Great
4. My Bad. I have updated the .gitignore file

The pull request is out: https://github.com/capriza/node-busmq/pull/23

The `busmq.js` and `busmq.min.js` have not been regenerated. Since we are at it, could you please

1. move these files to - say - a `dist` directory ?
2. have `browser-build.sh` merged in `package.json` at `scripts->build` or at least called by it.
This would make the project more friendly to IDEs such as IntelliJ/Webstorm by allowing exclusion of these files from source indexation, push buttons builds, etc...

**Comment by ronnen** [comment]
Looks like you've found a new friend on github:) does he know you're
practically blind these days?
If he annoys too much let me know. I have my ways...

Enjoy the lovely weekend,

Ronnen

On Mar 4, 2016, at 9:16 PM, Nadav Fischer <notifications@github.com> wrote:

I reviewed the code, it seems simple enough.

   1. I'm not a big fan of environment variables so I'll probably change
   that after I pull
   2. ioredis does not support the 'subscribe' event so need to emulate it,
   however you only subscribe to channel[0], but emit the subscribe event to
   all channels which isn;t very clean. I think it's safe to change the
   function to just handle the one channel instead of an array of channels so
   the code is tighter.
   3. For some time now I've been wanting to change the queue.push method
   to use the lua script to perform all of what the multi/exec does. This
   seems like a good opportunity so that there won't be any differences
   between the two libraries. I can do that after I pull.
   4. Please remove the dump.rdp file and the entire node_module/ directory
   you (probably accidently) committed.

Can you give it another go?

—
Reply to this email directly or view it on GitHub
<https://github.com/capriza/node-busmq/issues/22#issuecomment-192422732>.

**Comment by bgrieder** [comment]
@fujifish For sentinel/cluster  support I was thinking something around these lines,

`Bus.create({redis: [...]);`  -> current behaviour
`Bus.create({ioredis: {...});`  -> current behaviour with ioredis client
`Bus.create({sentinels: {...});`  -> sentinels support using ioredis
`Bus.create({sentinels: {...});`  -> cluster support using ioredis

wdyt?

**Comment by fujifish** [comment]
the `redis` option in the constructor is for specifying redis addresses in the form `redis://...`.
I think specifying a new `driver` or `engine` option would fit better (BTW, there is a `withRedis` method do to that, but I like `driver`/`engine` better)

for sentinels/clusters, how about we go with the same scheme as with setting the redis addresses, something like `redis: [sentinel://... sentinel://...]` and `redis: [cluster://..., cluster://...]`

Makes sense?

**Comment by bgrieder** [comment]
The ioredis parameters are rather inconsistent across direct connection/sentinels/cluster.
I appreciate you want to give some sense to this chaos yet care should be taken not to come up with too rigid a scheme that would prevent using certain potentially interesting options.

I have uploaded the list of options (Typescript syntax) that you can pass to ioredis for a direct connection (IORedisClientLayout), a sentinels setup (IORedisSentinelLayout) and a cluster setup (IORedisClusterLayout). 

As you can see for a direct connection for instance, an URL will not provide sufficient expressiveness to provide for all these options, even if, some do not make sense in the context of BusMQ.

Any further idea that could bring the best of both world is welcome.

[Configuration.txt](https://github.com/capriza/node-busmq/files/159876/Configuration.txt)

**Comment by fujifish** [comment]
I've merged your PR and made a few refactoring changes, nothing too dramatic.
I'm trying to get the tests to run sequentially for both node_redis and ioredis so any changes would be tested on both drivers.

I'd rather you open a new issue to add support for sentinels and clusters so we can track that separately.

Many thanks for this important contribution!

**Comment by fujifish** [comment]
@bgrieder I added that all tests now run both drivers to make sure nothing breaks when making further changes.
During testing I saw that the ioredis tests run a bit slower than node-redis tests, and I'm assuming that's because ioredis does not use pipelining by default as node-redis does. Just FYI.

**Comment by bgrieder** [comment]
Issue Created: https://github.com/capriza/node-busmq/issues/24

I will look at optimisations last.
